### ISSUE TITLE
New issue type: Detect spurious correlations for images

### ISSUE THREAD

**Issue by jwmueller** [cleanlab/cleanlab#860]
# Background

[This tutorial](https://docs.cleanlab.ai/master/tutorials/image.html) shows how Datalab can be applied to an image dataset with the auxiliary `image_key` argument, which will detect image-specific issues based on various scores from [CleanVision](https://github.com/cleanlab/cleanvision).

Each of these scores is essentially a basic function of the raw image quantifying basic properties like: how dark it is, how blurry, how large, etc.

# Objective 

The goal of the new issue type here is to measure the correlation between each of these individual image properties and the class labels (for an image classification dataset). If there is a strong correlation (i.e. predictive power) between one of these properties and the label, this is concerning and potentially indicative of a spurious correlation that is worth knowing about. 

# Steps to implement this

When this new `spurious_correlation` issue-type should be checked for:  If Datalab is being applied with the `image_key` argument specified. You should assume all of the CleanVision issue-types have already been detected by this code:
https://github.com/cleanlab/cleanlab/blob/master/cleanlab/datalab/internal/adapter/imagelab.py

and should avoid doing redundant computation (i.e. extract the image property scores from the imagelab code linked above, rather than writing new code to re-compute them).

After this, you should fit a univariate model to predict the labels separately from each value of the image-property. Here we want to use a multi-class classifier from sklearn that is computationally efficient and high-accuracy for univariate numeric data, options include:
- https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.GaussianNB.html#sklearn.naive_bayes.GaussianNB
- https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html#sklearn.neighbors.KNeighborsClassifier

To reiterate, the feature for this ML model will only have one column (one number extracted for each image, eg. the blurry score for say the blurry property).

You should use K-fold cross-validation to fit this classifier and get its out-of-sample class label predictions for every example in the dataset: https://docs.cleanlab.ai/stable/tutorials/pred_probs_cross_val.html

(class predictions will suffice, we don't need predicted class probabilities here). 

Finally, you should measure the (held-out) accuracy of these class predictions vs. the given class labels `A`. 

This issue will output a single global score `S` for the entire overall dataset.
Let `B` denote the baseline accuracy of simply always predicting the most common overall class label for every example in the dataset.

Then the score can simply be defined as: `S =  min(1 , (1-A)/(1-B+eps))`

This produces one score for the whole dataset, for each type of image property (dark score, blurry score, information score, size, aspect_ratio, etc.)

**Comment by 01PrathamS** [comment]
Hey @jwmueller  I have taken a look at this particular issue, and it seems to me that I might have the basic understanding required to address it 

 I would greatly appreciate your input on my initial approach please checkout, 

##########################################################################

import numpy as np
import pandas as pd
from sklearn.model_selection import cross_val_predict, cross_val_score
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

image_properties = ['dark_score', 'outlier_score', 'low_information_score', 'label_score', 'near_duplicate_score']
property_scores = {}
for property_of_interest in image_properties:
    X = image_properties_df[property_of_interest].values.reshape(-1, 1)
    y = image_properties_df['given_label'].values
    classifier = GaussianNB() 
    cv_predictions = cross_val_predict(classifier, X, y, cv=5)
    cv_accuracies = cross_val_score(classifier, X, y, cv=5, scoring='accuracy')
    mean_accuracy = np.mean(cv_accuracies)
    print(f'Accuracy for {property_of_interest}: {mean_accuracy}')


#########
# mean accuracy for dark_score: 0.14468333333333336
# mean accuracy for outlier_score: 0.1631333333333333
# mean accuracy for low_information_score: 0.2563666666666667
# mean accuracy for label_score: 0.16763333333333333
# mean accuracy for near_duplicate_score: 0.15618333333333334

**Comment by jwmueller** [comment]
@01PrathamS  nice you are on the right track! 

Some suggestions: 

- after you have the accuracy `A` as you've computed here, transform it to a score `S` as I outlined above.
- we only want to compute such scores for certain image properties where it we would be interested in a spurious correlation. So can you leave out `label_score` from your current list?  also the list should contain the rest of the issue types from [CleanVision](https://github.com/cleanlab/cleanvision/)
- For now, let's add this method into the `Datalab` class as an instance method: `datalab._spurious_correlation()`.  Note I suggest to keep it as a private method for now while we finalize the API later. For now, assume that the user has already called: `datalab.find_issues()` for each image property of interest, before they call: `datalab._spurious_correlation()`

Let me know if you have any questions; otherwise this looks like a good start!

**Comment by 01PrathamS** [comment]
Hello @jwmueller,I apologize for the delay in my submission.

please checkout, 

 import numpy as np
from sklearn.model_selection import cross_val_predict, cross_val_score
from sklearn.naive_bayes import GaussianNB
import warnings
from statistics import mode

def _spurious_correlation(self):
        image_properties_df = lab.get_issues()
        baseline_accuracy = data["label"].count(mode(data["label"])) / len(data["label"])
        eps = 1e-8
        image_properties = ["near_duplicate_score", "blurry_score", "light_score", "low_information_score", "dark_score", "grayscale_score", "odd_aspect_ratio_score", "odd_size_score"]
        property_scores = {}
        for property_of_interest in image_properties:
            X = image_properties_df[property_of_interest].values.reshape(-1, 1)
            y = image_properties_df["label"]
            classifier = GaussianNB()
            cv_predictions = cross_val_predict(classifier, X, y, cv=5)
            cv_accuracies = cross_val_score(classifier, X, y, cv=5, scoring='accuracy')
            mean_accuracy = np.mean(cv_accuracies)
            S = min(1, (1 - mean_accuracy) / (1 - baseline_accuracy + eps))
            print(f'Overall Datascore for {property_of_interest}: {S}')

## image_properties_df by calling >> lab.get_issues()
# 	is_label_issue	label_score	is_outlier_issue	outlier_score	is_near_duplicate_issue	near_duplicate_score	is_non_iid_issue	non_iid_score	is_dark_issue	odd_size_score	...	is_light_issue	dark_score	odd_aspect_ratio_score	grayscale_score	is_grayscale_issue	is_low_information_issue	light_score	low_information_score	is_odd_size_issue	is_blurry_issue
# 0	True	0.163778	False	0.957810	False	0.038207	False	0.957125	False	1.0	...	False	1.000000	1.0	0	True	False	1.0	0.326907	False	False
# 1	False	0.972319	False	0.982078	False	0.012460	False	0.924408	False	1.0	...	False	0.869333	1.0	0	True	False	1.0	0.474813	False	False
# 2	False	0.997889	False	0.959157	False	0.034223	False	0.955660	False	1.0	...	False	0.901961	1.0	0	True	False	1.0	0.354166	False	False
# 3	False	0.929160	False	0.955932	False	0.037803	False	0.893504	False	1.0	...	False	0.777137	1.0	0	True	False	1.0	0.296907	False	False
# 4	False	0.998787	False	0.985570	False	0.011249	False	0.891352	False	1.0	...	False	0.924902	1.0	0	True	False	1.0	0.377497	False	False


# Overall Datascore for near_duplicate_score: 0.9375740636565844
# Overall Datascore for blurry_score: 0.8307222129919754
# Overall Datascore for light_score: 0.999999988888889
# Overall Datascore for low_information_score: 0.8262592500786009
# Overall Datascore for dark_score: 0.9503518412923868
# Overall Datascore for grayscale_score: 0.999999988888889
# Overall Datascore for odd_aspect_ratio_score: 0.999999988888889
# Overall Datascore for odd_size_score: 0.999999988888889


if this finds well, i add this method into the Datalab class (cleanlab\cleanlab\datalab\datalab.py) as an instance method right!!

**Comment by jwmueller** [comment]
yes for now my suggestion is to add it to [datalab.py](https://github.com/cleanlab/cleanlab/blob/master/cleanlab/datalab/datalab.py) as a PRIVATE instance method called say:  `_spurious_correlations()`

Your method can assume the user has already called: `datalab.find_issues(..., image_key=...)`
so that all the cleanvision stuff has already been run by the time your method is running.

Try to keep the code for your new method short and rely on helper functions. 
The helper functions can all live in a new file:  cleanlab/datalab/internal/spurious_correlation.py

After your first PR is in, we can revisit how this code fits into the typical Datalab run, and improve the user experience. That part should be super easy if the initial implementation is solid.

**Comment by jwmueller** [comment]
@01PrathamS  your code looks good enough for a PR! will leave my feedback in the PR. 
Main suggestion for the PR is to include an end-to-end unit test of this functionality. You should actually create a toy dataset that suffers from a spurious correlation (say have 10 tiny images at varying levels of darkness, and make the label related to how dark they are). And then verify that this code detects this spurious correlation. Likewise your. same unit test should verify that the other spurious correlation scores (those unrelated to `dark`, `light`) do NOT give low scores for this same dataset.

**Comment by jwmueller** [comment]
added in https://github.com/cleanlab/cleanlab/pull/1140
